src/addons/messagelog/src/main/java/ee/ria/xroad/proxy/messagelog/AbstractTimestampRequest.java

```diff
@@ -62,14 +62,20 @@
         TimeStampRequest tsRequest = createTimestampRequest(getRequestData());
 
         TsRequest req = makeTsRequest(tsRequest, tspUrls);
-        if (req.getInputStream() == null) {
-            throw new RuntimeException("Could not get response from TSP");
-        }
 
         TimeStampResponse tsResponse = getTimestampResponse(req.getInputStream());
         log.info("tsresponse {}", tsResponse);
-        verify(tsRequest, tsResponse);
 
+        try {
+            verify(tsRequest, tsResponse);
+
+        } catch (Exception ex) {
+            // Exceptions caused by invalid TSA responses are catched here.
+            log.error("Failed to verify time stamp from " + req.getUrl(), ex);
+            log.debug("Remove {} from tspUrls and try again", req.getUrl());
+            tspUrls.remove(req.getUrl());
+            execute(tspUrls);
```

Comment:
- consider moving the loop (or all logic) in makeTsRequest to this function instead of recursively calling execute.

Reply:
Moved all request logic in `makeTsRequest` method.